graphics/fonts/macfont.cpp

```diff
@@ -581,54 +574,31 @@ MacFONTFont *MacFONTFont::scaleFont(const MacFONTFont *src, int newSize, int sla
 
 	int dstPitch = data._rowWords;
 
+	Graphics::Surface tmpSurf;
+	int dstGraySize = newSize * 20 * newSize;
+	int *dstGray = 0;
+	tmpSurf.create(MAX(src->getFontSize() * 2, newSize * 2), MAX(src->getFontSize() * 2 + 2, newSize * 2 + 2),
+		PixelFormat::createFormatCLUT8());
+
 	for (uint i = 0; i < src->_data._glyphs.size() + 1; i++) {
 		const MacGlyph *srcglyph = (i == src->_data._glyphs.size()) ? &src->_data._defaultChar : &src->_data._glyphs[i];
 
 		int grayLevel = src->_data._fRectHeight * srcglyph->width / 4;
 
 #if DEBUGSCALING
 		int ccc = 'c';
-		dododo = i == ccc;
 #endif
 
-		srcSurf.fillRect(Common::Rect(srcSurf.w, srcSurf.h), 0);
-		src->drawChar(&srcSurf, i + src->_data._firstChar, 0, 0, 1);
-		memset(dstGray, 0, dstGraySize * sizeof(int));
-		magnifyGray(&srcSurf, dstGray, srcglyph->width, src->_data._fRectHeight, scale);
-
 		MacGlyph *glyph = (i == src->_data._glyphs.size()) ? &data._defaultChar : &data._glyphs[i];
-		int *grayPtr = dstGray;
-
-		for (int y = 0; y < data._fRectHeight; y++) {
-			byte *dst = (byte *)srcSurf.getBasePtr(0, y);
-
-			for (int x = 0; x < glyph->bitmapWidth; x++, grayPtr++, dst++) {
-#if DEBUGSCALING
-				if (i == ccc) {
-					if (*grayPtr)
-						debugN(1, "%3d ", *grayPtr);
-					else
-						debugN(1, "    ");
-				}
-#endif
-				if (*grayPtr > grayLevel)
-					*dst = 1;
-				else
-					*dst = 0;
-			}
-#if DEBUGSCALING
-			if (i == ccc)
-				debug(1, "");
-#endif
-		}
-
+		Graphics::Surface *srcSurf = scaleFont->scaleSingleGlyph(glyph->bitmapWidth, data._fRectHeight, grayLevel, i + src->_data._firstChar,
+																 0, src->_data._fRectHeight, srcglyph->width, scale, &dstGray);
 		if (slant & kMacFontBold) {
 			memset(dstGray, 0, dstGraySize * sizeof(int));
```

Comment:
How do you know that `dstGray` now is of `dstGraySize`?

Reply:
yes, so dstGray was originally the size of dstGraySize, I have declared these above on line 578 and 579.

Reply:
It is being set in the intializeScaling function to same size as newsize * 20 * newsize

Reply:
What will happen if a developer accidentally changes it to something different in the initializeScaling()?